[Mobile Payments] Present incomplete store address error to user

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes #5088
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This change will present an error to the user when they need to update their store address in order to connect to a reader. In a further PR, it will include a button with a link to the WP-Admin page where they can make the change.

<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
1. With a store set up for In-Person Payments, go to `WP-Admin > WooCommerce > Settings` and delete the contents of `Address line 1`
2. In the app, go to `⚙️ > In-Person Payments > Manage Card Reader`
3. Search for and connect to a reader (or simulated reader)
4. Observe that the error message is shown telling the merchant to update their address.
5. Update the address in WP-Admin to include `Address line 1` again
6. Tap retry, and connect to the reader again.

Unit tests will follow in a future PR. The retry action could use some polish; e.g. just retrying connection to the exact reader which failed would be much nicer than starting a search again... but that can be an iteration on this for after we send the user to the `adminUrl` or something.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://user-images.githubusercontent.com/2472348/143595235-441c848e-44e8-41c9-a1da-0646599855bb.mp4


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

joshheald commented:
@ctarda I'd appreciate your thoughts on this as I'm starting to agonise over this. In particular, do you think I'm putting too much into `UnderlyingError+Stripe.swift` with it knowing about `DotcomError` and depending on `Networking`. 

I'm not a massive fan of that dependency, but it does seem like the best place for us to figure out that we've got this error, if we want to present it wherever we need to.

joshheald commented:
Hmm... tests on CI aren't happy about the dependency either!

ctarda commented on Hardware/Hardware/CardReader/StripeCardReader/UnderlyingError+Stripe.swift:
```diff
@@ -1,9 +1,18 @@
 import StripeTerminal
+import Networking
 /// Maps Stripe SDK specific errors to domain errors:
```
@joshheald As you mentioned in the comments, it seems like we might mixing concerns a little bit, with leads to tangle the dependency tree a little bit as well.

It seems to me like what we want is to pass up to the UI an error that is not actually thrown by the Terminal SDK, but that happens when a network request is failing. 

In one sentence: we want to pass an error to the user when a network request is failing. That request is the one that fetches the location, and is implemented by WCPayRemote.

That request fetches the location id, and what we want is to:
1. Stop the attempt to connect to the reader
2. Show a user facing error.

To me, the key is point #1. We want to stop the connection to a reader when we can not get a location identifier. That, to me, is business logic.

That's all to say that I think we should move the logic out of Hardware (which doesn't do any other networking) at least one layer up.

UnderlyigError conforms to Error, so we could do a bit of good old fashioned polymorphism. WCPayTokenProvider is the entity that encapsulates the business logic of fetching the location id. At this point, it is passing up whatever error is coming from the networking layer. 

I would explore declaring a new error alongside it (it could be an inner type to WCPayTokenProvider or it could be just a public entity declared next to it), so that WCPayTokenProvider.fetchDefaultLocationID actually throws one instance of this new error.

If everything else works as expected, this new error can be passed to CardReaderServiceError as an underlying error in when captured in StripeCardReaderService.getBluetoothConfiguration (this might or might not require adding a new case to CardReaderServiceError)

From that on, the error should be propagated up to the UI without other intervention. It might be necessary to add some logic to the UI as well to make sure that the payment intent is cancelled if connection to the reader fails.

But, in a nutshell, I would explore to:
1. Declare a new error type that would be initialized by WCPayTokenProvider when the request to get the location id fails.
2. Pass that new error in the completion block of fetchDefaultLocationID, instead of the Dotcom error that is coming from the remote service.
3. Take it from there 😄

joshheald committed:
5088 Parse incompleteStoreAddress cardreader error

joshheald commented on Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift:
```diff
@@ -281,13 +281,25 @@ extension StripeCardReaderService: CardReaderService {
         }
     }
 
+    private func underlyingError(from rawError: LocalizedError) -> LocalizedError {
```
I'm leaving this slightly confusingly named function (that doesn't necessarily make an `UnderlyingError`) because I think that the next steps are to make `UnderlyingError` an abstract parent of both `StripeUnderlyingError` (or something) and `CardReaderConfigError`

ctarda replied:
I would prefer to stay away from creating a hierarchy of errors. All of them implement Error after all, so conforming to Error seems like is all we might need.

I do agree this function name could be less confusing. What about extracting it to a factory in UnderlyingError? Something like

```
extension UnderlyingError {
   init(with configurationError: CardReaderConfigurationError) {
   }
}
```

That way, I think, we could also avoid the downcast to CardReaderConfigError as well.

ctarda replied:
Actually, I am not sure if it's actually possible to type the parameter as CardReaderConfigurationError, so the downcast might not be possible to avoid

joshheald replied:
Yeah I tried to make the hierarchy and quickly noped out of that, so I'm glad you don't love that idea!

I'll work on the naming. I could maybe remove this method by using a failable initializer in the extension you suggest.

joshheald committed:
WIP

ctarda committed:
Make CardReaderConnectionController build

ctarda committed:
Expose a property in the underlying error to indicate if it's a location id missing

ctarda committed:
Rely on UnerlyingError's factory only

joshheald commented on Hardware/Hardware/CardReader/CardReaderServiceError.swift:
```diff
@@ -192,6 +192,19 @@ public enum UnderlyingError: Error {
     /// internal state of the CardReaderService.
     case internalServiceError
+
+    case locationIdMissing
```
I'd read `UnderlyingError` as being more bound to Stripe, but really that's just the extension with Stripe in the name, so adding this makes sense and avoids some of the polymorphism gymnastics I was having to do. Thanks!

joshheald replied:
I think this could do with also being called `incompleteStoreAddress`, since we want to surface that error to the merchant, and will need to include the `adminUrl` associated value as well.

ctarda replied:
Absolutely! I am terrible at naming things

joshheald commented on Hardware/Hardware/CardReader/CardReaderServiceError.swift:
```diff
@@ -192,6 +192,19 @@ public enum UnderlyingError: Error {
+        case .incompleteStoreAddress(_):
+            self =  .locationIdMissing
+        case .unknown(let error):
+            self  = UnderlyingError(with: error)
```
It's convenient to fall back to the `+Stripe` extension here, which provides default behaviour of mapping errors it can't parse to `.internalServerError`... but it does seem like a bit of a side effect.

Two options come to mind:
1. Make this a failable init, and then at the call site use the Stripe init where it's appropriate.
2. Add a `.unknown` case to the `UnderlyingError` and use that, rather than always falling back to `.internalServerError`

I'm going to try option 1, because I think it'll be more explicit and help us see any consequences of the change.

ctarda replied:
I would lean more towards option 2, to be honest. The way we would not need to deal with special cases at the call site

joshheald replied:
To follow up from a slack chat; I gave the `.unknown` case a go here and it didn't really add anything. `.internalServiceError` is basically the unknown case already, it's intended as a catch all, and that's how we use it. 

It is quite close to internalServ**er**Error, which might lead to some confusion, so we could consider renaming it.

joshheald commented on Hardware/Hardware/CardReader/CardReaderServiceError.swift:
```diff
@@ -209,6 +222,10 @@ extension UnderlyingError {
         }
     }
+
+    public var isLocationIDMissingError: Bool {
```
Not totally sure we need this for a single case? We might need to add `.storePostcodeMissing` for #5320, in which case it could be helpful, but I could also see that error being collapsed into this one. Regardless, we'll need to unwrap and downcast at the call site so that we can get the `adminUrl` out when that's included

ctarda replied:
Maybe a more generic name like `isLocationError`? It seems like marking these special cases clearly simplifies the client code a lot.

joshheald commented on WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift:
```diff
@@ -518,7 +518,13 @@ private extension CardReaderConnectionController {
-        showConnectionFailed(error: error)
+
+        if case CardReaderServiceError.connection(underlyingError: let underlyingError) = error, underlyingError.isLocationIDMissingError {
+            return onConnectionFailed(error: error)
```
I used that name first too, and didn't like it because it's so similar to the enclosing method name. 

Did you want to inline the error creation here for a particular reason? I quite liked extracting it to another method which then differentiated between the default alert and the missing address alert. I'm going to go back to that for now, and try to name stuff better, as the previous code was handling both alert cases.

ctarda replied:
I have no preference here, to be honest. This looked slightly less verbose to me, and that's why I went with it 😄

joshheald commented on WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift:
```diff
@@ -562,12 +568,9 @@ private extension CardReaderConnectionController {
-            alerts.connectingFailed(from: from, continueSearch: continueSearch, cancelSearch: cancelSearch)
-        }
+        // This is the alert specific to a failure in the connection due to a missing location id.
+        // Might be worth passing the store URL?
```
Yeah, agreed, planning to do that in a later commit/PR

joshheald reviewed (commented):
@ctarda thanks for doing so much on it! I've added a review as a more convenient way for us to discuss, and as notes for myself as I work on these changes. Chime in with any thoughts 😊

joshheald committed:
5088 Include card config errors in UnderlyingError

joshheald committed:
5088 Create UnderlyingError from multiple types

joshheald committed:
Update mock to include missingAddress function

joshheald committed:
5088 Include retry button for after address update

joshheald committed:
5088 Retry reader connection from scratch

Previously, we continued the search when the user tapped `Try Again`. Unfortunately, this would never find the reader they had just failed to connect to, so it's not really trying again.

This change resets the search and re-runs it, so the original reader can be found again. There may be better ways to do this.

allendav commented on WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalConnectionFailedUpdateAddress.swift:
```diff
@@ -0,0 +1,65 @@
+private extension CardPresentModalConnectingFailedUpdateAddress {
+    enum Localization {
+        static let title = NSLocalizedString(
+            "Please update your store address to proceed.",
```
Could we be more specific? i.e. "update" seems vague. Could we use the word "correct" or "fully set" or something instead? cc @adamzelinski

allendav commented on WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift:
```diff
@@ -561,6 +565,31 @@ private extension CardReaderConnectionController {
+
+        switch underlyingError {
+        case .incompleteStoreAddress(adminUrl: _):
+            alerts.connectingFailedMissingAddress(from: from, continueSearch: continueSearch, cancelSearch: cancelSearch)
```
Instead of MissingAddress, would it be more appropriate to say InvalidStoreAddress or BadStoreAddress?

allendav replied:
In light of #5560, which I noticed after I reviewed this PR, I'd opt for "Incomplete" instead of "Missing"

joshheald replied:
Because there's a string of PRs to merge here, to avoid messy merges I'm going to open issues for minor changes and address them next, after merging all PRs. If there's anything fundamental, I won't merge until fixing, but from a skim it doesn't look like there is. See #5567

allendav commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlerts.swift:
```diff
@@ -25,6 +25,12 @@ final class CardReaderSettingsAlerts: CardReaderSettingsAlertsProvider {
+    func connectingFailedMissingAddress(from: UIViewController,
+                                        continueSearch: @escaping () -> Void,
+                                        cancelSearch: @escaping () -> Void) {
+        setViewModelAndPresent(from: from, viewModel: connectingFailedUpdateAddress(continueSearch: continueSearch, cancelSearch: cancelSearch))
```
Instead of continue, since it looks like we need to "retry" it, could we say "retrySearch" ?

joshheald replied:
This is done in #5552 and #5560

allendav commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlerts.swift:
```diff
@@ -160,6 +166,10 @@ private extension CardReaderSettingsAlerts {
         CardPresentModalConnectingFailed(continueSearch: continueSearch, cancelSearch: cancelSearch)
     }
 
+    func connectingFailedUpdateAddress(continueSearch: @escaping () -> Void, cancelSearch: @escaping () -> Void) -> CardPresentPaymentsModalViewModel {
```
Same comment as above

allendav commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsAlertsProvider.swift:
```diff
@@ -45,6 +45,13 @@ protocol CardReaderSettingsAlertsProvider {
+    /// Defines an alert indicating connecting failed because their address needs updating.
+    /// The user may try again or cancel
+    ///
+    func connectingFailedMissingAddress(from: UIViewController,
```
Instead of MissingAddress, would it be more appropriate to say InvalidStoreAddress or BadStoreAddress?

allendav replied:
In light of #5560, which I noticed after I reviewed this PR, I'd opt for "Incomplete" instead of "Missing"

allendav commented on WooCommerce/WooCommerceTests/Mocks/MockCardReaderSettingsAlerts.swift:
```diff
@@ -83,6 +83,18 @@ final class MockCardReaderSettingsAlerts: CardReaderSettingsAlertsProvider {
     }
 
+    func connectingFailedMissingAddress(from: UIViewController,
+                                        continueSearch: @escaping () -> Void,
```
Same comments as above re missing and continue

allendav commented on Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:
```diff
@@ -242,26 +242,42 @@ private final class WCPayTokenProvider: CardReaderConfigProvider {
     }
 
-    func fetchToken(completion: @escaping(Result<String, Error>) -> Void) {
+    func fetchToken(completion: @escaping(Result<String, CardReaderConfigError>) -> Void) {
```
Do we need a separate error (CardReaderConfigError) apart from the CardReaderServiceError enumeration you've added for bad address above?

Since the CardPresentPaymentStore manages all access to the CardReaderService, it would seem appropriate to leave the function signatures as Error and construct and return a CardReaderService.incompleteStoreAddress in the event the DotcomError has store_address_is_incomplete, but otherwise forward the Error intact from the remote?

joshheald replied:
Well... I don't really have strong feelings on it. We don't _need_ the separate error, in that I've not used it and there's no specific handling of the `.unknown(_)` case. So it would be trivial to retain the `Result<String, Error>` signatures of these two methods and I'm happy to do that.

I added the `.unknown` case because I think it's useful to give some context of what the error is about in the type, while still providing the detail of the error for downstream use. We can just apply YAGNI here and I'll do that in the rework branch from #5567, and see what it looks like there.

allendav reviewed (commented):
I'm curious of your thoughts, especially on https://github.com/woocommerce/woocommerce-ios/pull/5515/files#r759598494

allendav reviewed (approved):
We can defer `ReaderTokenProvider` and `ReaderLocationProvider` and `CardReaderConfigProvider` tuple change discussions to a separate issue.
